tmt/steps/provision/__init__.py

```diff
@@ -472,6 +477,7 @@ class GuestSshData(GuestData):
     key: List[str] = dataclasses.field(default_factory=list)
     # password
     password: Optional[str] = None
+    ssh_option: Tuple[str, ...] = dataclasses.field(default_factory=tuple)
```

Comment:
Why are we using `Tuple` here and `List` for the `key` attribute? Aren't they almost the same? If we use `List` for ssh options as well we would not need casting in the provision plugins, right?

Reply:
This smells like something between Click and fmf, let's see.

Unfortunately, resolving the question of different types will not help us get rid of `cast()` calls in provision plugins.

```python
    # FIXME: cast() - typeless "dispatcher" method
    data.ssh_option = cast(Tuple[str, ...], self.get('ssh-option'))
```

Even if we unify the type to be always `List[str]`, the cast would just switch from `Tuple` to said `List`, because the issue lies in `self.opt()`. As hinted in the comment, `self.opt()` returns `Any` as it's not aware of type annotations, and there's no (easy) way how to propagate the fact that the type of its return value depends on the value of its input.

I plan to fix this, somehow, and I tend to add `cast()` calls to annotate whatever `opt()` or `get()` return when assigning it to local names.

Reply:
I'm afraid we're out of luck, with `multiple=True`, Click will return collected options in a tuple, even if the `default` was an empty list (`default=[]`).

This means we will need to coerce one value, either

* the one from `self.opt()`, to `List`, to match what we might read from fmf data one day, or
* the one from fmf, in the future `_normalize_ssh_option`, to `Tuple`, matching type from Click.

Since guest data is stored in YAML, I'd probably stay away from tuples completely, and coerce the value from `self.opt()` to list.

Reply:
Trying in [1bd2054](https://github.com/teemtee/tmt/pull/1673/commits/1bd20547e7ff5f3b2d13318573455c901a220d2c)

Reply:
Thanks for the detailed explanation. Let's merge the amazing #1558 in `1.20`! :-)